core/domain/prod_validation_jobs_one_off.py

```diff
@@ -110,67 +351,3183 @@ def validate(cls, item):
+        """Validate that sender email corresponds to email of user obtained
+        by using the sender_id.
+
+        Args:
+            item: SentEmailModel to validate.
+        """
+        _, sender_model_tuples_list = cls.external_models['sender_id']
+
+        for (_, sender_model) in sender_model_tuples_list:
+            if sender_model and not sender_model.deleted:
+                if sender_model.email != item.sender_email:
+                    cls.errors['sender email check'].append((
+                        'Entity id %s: Sender email %s in entity does not '
+                        'match with email %s of user obtained through '
+                        'sender id') % (
+                            item.id, item.sender_email, sender_model.email))
+
+    @classmethod
+    def _validate_recipient_email(cls, item):
+        """Validate that recipient email corresponds to email of user obtained
+        by using the recipient_id.
+
+        Args:
+            item: SentEmailModel to validate.
+        """
+        _, recipient_model_tuples_list = cls.external_models['recipient_id']
+
+        for (_, recipient_model) in recipient_model_tuples_list:
+            if recipient_model and not recipient_model.deleted:
+                if recipient_model.email != item.recipient_email:
+                    cls.errors['recipient email check'].append((
+                        'Entity id %s: Recipient email %s in entity does '
+                        'not match with email %s of user obtained through '
+                        'recipient id') % (
+                            item.id, item.recipient_email,
+                            recipient_model.email))
+
+    @classmethod
+    def _get_custom_validation_functions(cls):
+        return [
+            cls._validate_sent_datetime, cls._validate_sender_email,
+            cls._validate_recipient_email]
+
+
+class BulkEmailModelValidator(BaseModelValidator):
+    """Class for validating BulkEmailModels."""
+
+    @classmethod
+    def _get_model_id_regex(cls, item):
+        return '.'
+
+    @classmethod
+    def _get_json_properties_schema(cls, item):
+        return {}
+
+    @classmethod
+    def _get_model_domain_object_instances(cls, item):
+        return []
+
+    @classmethod
+    def _get_commit_cmd_domain_class(cls):
+        return None
+
+    @classmethod
+    def _get_external_id_relationships(cls, item):
+        return {
+            'recipient_id': (
+                user_models.UserSettingsModel, item.recipient_ids),
+            'sender_id': (user_models.UserSettingsModel, [item.sender_id]),
+        }
+
+    @classmethod
+    def _validate_id_length(cls, item):
+        """Validate that model id has length 12.
+
+        Args:
+            item: BulkEmailModel to validate.
+        """
+        # The bulk email model has ids as randomly generated strings of
+        # length 12.
+        if len(item.id) != email_models.BulkEmailModel.MODEL_ID_LENGTH:
```

Comment:
Surely this can be checked via the id regex.

Reply:
Updated.